### PR TITLE
Created Broadcast Plugin

### LINKED ISSUE #271: !broadcast class_channels
Broadcast message to all the class channels (or select channels) from the admin_chat channel. 

I'd like to send reminders to all the class channels about specific events without having to go through each channel and paste the message.  It doesn't need to be a constant reminder like with politics, just whenever an admin needs to send a message to all/select courses.

### PR DESCRIPTION
**Pull request by tanndlin**
image
image
fixes #271

### CODE REVIEW

File: src/app/plugins/broadcast.plugin.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+export class BroadcastPlugin extends Plugin {
+  public name: string = 'Broadcast';
+  public description: string = 'Sends an annoucement to all class channels';
+  public usage: string = '!broadcast <message>';
+  public pluginAlias = [];
+  public permission: ChannelType = ChannelType.Admin;
+  public pluginChannelName: string = Constants.Channels.Staff.AdminChat;
+
+  private _LION_PFP_URL: string =
+    'https://cdn.discordapp.com/avatars/574623716638720000/7d404c72a6fccb4a3bc610490f8d7b72.png';
```
Comment (c650) [inline src/app/plugins/broadcast.plugin.ts:15]: is it worth making this a global constant

File: src/app/plugins/broadcast.plugin.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    if (args[0].toLowerCase() === 'confirm') {
+      message.reply(
+        `Sending Annoucement to \`${classChans.length}\` classes... I will let you know it has finished`
+      );
+
+      const promises = classChans.map((chan) => (chan as TextChannel).send(this._STATE));
+      this._STATE = undefined;
+
+      Promise.all(promises).then(() => message.reply('Annoucement sent!'));
```
Comment (c650) [inline src/app/plugins/broadcast.plugin.ts:56]: You should be able to await the `Promise.all`

File: src/app/plugins/broadcast.plugin.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import Constants from '../../common/constants';
+import { RichEmbed, TextChannel } from 'discord.js';
+
+export class BroadcastPlugin extends Plugin {
+  public name: string = 'Broadcast';
+  public description: string = 'Sends an annoucement to all class channels';
+  public usage: string = '!broadcast <message>';
+  public pluginAlias = [];
+  public permission: ChannelType = ChannelType.Admin;
+  public pluginChannelName: string = Constants.Channels.Staff.AdminChat;
```
Comment (c650) [inline src/app/plugins/broadcast.plugin.ts:12]: should be necessary https://github.com/joey-colon/lion/blob/5d686fd2ead1da4121b2e2348e5401c5da0d683d/src/app/plugins/addclasschans.plugin.ts#L15

File: src/common/constants.ts [hunk 4]
```diff
@@ -41,6 +41,7 @@ const Constants = {
     },
     Staff: {
       ModChat: 'mod_chat',
+      AdminChat: 'admin_chat',
```
Comment (c650) [inline src/common/constants.ts:44]: not necessary https://github.com/joey-colon/lion/blob/5d686fd2ead1da4121b2e2348e5401c5da0d683d/src/app/plugins/addclasschans.plugin.ts#L15

**Comment by c650** [comment]
Should we also let the caller specify a subset of class channels

File: src/app/plugins/broadcast.plugin.ts (hunk 1 above)
Comment (tanndlin) [reply src/app/plugins/broadcast.plugin.ts:15]: Probably

Comment (tanndlin) [reply src/app/plugins/broadcast.plugin.ts:15]: Should I put it in Constants.ts

Comment (c650) [reply src/app/plugins/broadcast.plugin.ts:15]: yes

**Comment by tanndlin** [comment]
> Should we also let the caller specify a subset of class channels

Subse meaning IT CS etc?

File: src/app/plugins/broadcast.plugin.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+      }
+
+      //Otherwise, its a class name
+      const targetChannel = this.container.guildService.getChannel(name.toLowerCase());
+      if (targetChannel) {
+        this._CHANS_TO_SEND.push(targetChannel);
+      }
+    });
+
+    //Remove possible duplicates from list
```
Comment (c650) [inline src/app/plugins/broadcast.plugin.ts:99]: nit space after //

File: src/app/plugins/broadcast.plugin.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    this._CHANS_TO_SEND = [...new Set(this._CHANS_TO_SEND)];
+    this._reportToUser(message);
+  }
+
+  private _reportToUser(message: IMessage) {
+    message.reply(
+      `You are about to send:\n\`\`\`${this._EMBED_TO_SEND?.description}\`\`\`\n to \`${this._CHANS_TO_SEND.length}\` classes... Are you sure?\nRespond with \`confirm\` or \`cancel\``
```
Comment (c650) [inline src/app/plugins/broadcast.plugin.ts:106]: break this up so it's not so wide

'msg1'
+ 'msg2'
+ 'msg3'

**Review by c650 — APPROVED** [review]

**Review by c650 — APPROVED** [review]

**Review by joey-colon — APPROVED** [review]

**Pull request merged**